File: code/betanmf_timed.m
% This function decomposes input matrix by nonnegative matrix factorization
% (NMF) based on beta-divergence criterion and multiplicative update rules.
%
% The input nonnegative data matrix X will be decomposed as X = A*B
% approximately, where matrices A and B are often called "basis matrix" and
% "activation matrix," respectively. All the entries in X, A, and B must be
% nonnegative values. All the update rules implemented in this function are
% based on majorization-minimization algorithm.
%
% See also:
%
%   C. Fevotte, J. Idier, "Algorithms for nonnegative matrix factorization
%   with the beta-divergence."
%
%   M. Nakano, H. Kameoka, J. Le Roux, Y. Kitano, N. Ono, and S. Sagayama,
%   "Convergence-guaranteed multiplicative algorithms for nonnegative matrix
%   factorization with beta-divergence," Proc. MLSP, pp. 283-288, 2010.
%
%   http://d-kitamura.net
%
% Coded by D. Kitamura (d-kitamura@ieee.org) on 31 Jul, 2015 (ver1.0).
% Modified by D. Kitamura (d-kitamura@ieee.org) on 18 Nov, 2018 (ver1.1).
%
function [A, B, f, time_A, time_B, time_cost] = betanmf_timed (X, A, B, tol, maxiter, verbose)

  % Set zeros to small positive numbers to prevent numerical issues in
  % the updates below.
  [n p] = size(X);
  X     = max(X,eps);
  E     = ones(n,p);

  % Handle optional arguments.
  if nargin < 4
    tol = 1e-6;
  end
  if nargin < 5
    maxiter = 5000;
  end 
  if nargin < 6
    verbose = true;
  end
  
  n_steps_A = 7;
  n_steps_B = 7;
  time_A = zeros(maxiter, n_steps_A);
  time_B = zeros(maxiter, n_steps_B);
  time_cost = zeros(maxiter, 1);
  

  % Compute the value of the objective function at the initial estimate
  % of the solution.
  f    = zeros(maxiter + 1,1);
  f(1) = cost(X,A*B);
  if verbose
    fprintf('iter objective (cost fn) max.diff\n');
    fprintf('---- ------------------- --------\n');
  end

  % Repeat until the maximum number of iteration is reached, or until the
  % convergence criterion is met.
  for i = 1:maxiter
    A0     = A;
    B0     = B;
    [A B timeA timeB]  = update_timed(X,A,B,E);
    time_A(i,:) = timeA;
    time_B(i,:) = timeB;
    
    tic;
    f(i+1) = cost(X,A*B);
    time_cost(i) = toc;
    d      = max(max(max(abs(A - A0))),...
                 max(max(abs(B - B0))));
    if verbose
      fprintf('%4d %+0.12e %0.2e\n',i,f(i+1),d);
    end
    if d < tol
      break
    end
  end
  f = f(1:i+1);

  
function [A,B, timeA, timeB] = update_timed(X,A,B,E)
    [A, timeA] = update_timed_A(X, A, B, E);
    [B, timeB] = update_timed_B(X, A, B, E);

  
function [A,timeA] = update_timed_A(X, A, B, E)
    timeA = [];
%   A = A .* (((X ./ (A*B)) * B.') ./ (E * B.'));
%   A = max(A,eps);
%   B = B .* ((A.' * (X ./ (A*B))) ./ (A.' * E));
%   B = max(B,eps);
    % step 1
    tic;
    M = A*B;
    timeA = [timeA, [toc]];
    
    % step 2
    tic;
    M = X ./ M;
    timeA = [timeA, [toc]];
    
    % step 3
    tic;
    M = M * B';
    timeA = [timeA, [toc]];
    
    % step 4
    tic;
    D = E * B';
    timeA = [timeA, [toc]];
    
    % step 5
    tic;
    M = M ./ D;
    timeA = [timeA, [toc]];
    
    % step 6
    tic;
    A = A .* M;
    timeA = [timeA, [toc]];
    
    % step 7
    tic;
    A = max(A,eps);
    timeA = [timeA, [toc]];
    

function [B,timeB] = update_timed_B(X, A, B, E)
    timeB = [];
%   A = A .* (((X ./ (A*B)) * B.') ./ (E * B.'));
%   A = max(A,eps);
%   B = B .* ((A.' * (X ./ (A*B))) ./ (A.' * E));
%   B = max(B,eps);
    % step 1
    tic;
    M = A*B;
    timeB = [timeB, [toc]];
    
    % step 2
    tic;
    M = X ./ M;
    timeB = [timeB, [toc]];
    
    % step 3
    tic;
    M = A' * M;
    timeB = [timeB, [toc]];
    
    % step 4
    tic;
    D = A' * E;
    timeB = [timeB, [toc]];
    
    % step 5
    tic;
    M = M ./ D;
    timeB = [timeB, [toc]];
    
    % step 6
    tic;
    B = B .* M;
    timeB = [timeB, [toc]];
    
    % step 7
    tic;
    B = max(B,eps);
    timeB = [timeB, [toc]];
    
    





% This implements the multiplicative updates.
% function [A, B] = update (X, A, B, E)
%   A = A .* (((X ./ (A*B)) * B.') ./ (E * B.'));
%   A = max(A,eps);
%   B = B .* ((A.' * (X ./ (A*B))) ./ (A.' * E));
%   B = max(B,eps);

File: code/betanmf_timed.R
# This function decomposes the input matrix X = A*B by nonnegative
# matrix factorization (NMF) based on the beta-divergence criterion
# (negative Poisson log-likelihood) and multiplicative update
# rules. All entries of initial estimates A and B should be
# positive. This is adapted from the MATLAB code by D. Kitamura
# (http://d-kitamura.net).

cost <- function(X,AB, e){
        return(sum(AB - X * log(AB + e)))
}

update_A_timed <- function(X,A,B,E,e){
  # original update is:
  #A <- A * (((X / (A %*% B)) %*% t(B)) / (E %*% t(B)))
  #A <- pmax(A,e)
  time = c()

  ## step 1
  start = proc.time()
  M = A %*% B
  runtime = proc.time() - start
  time = c(time, runtime[[3]])

  ## step 2
  start = proc.time()
  M = X / M
  runtime = proc.time() - start
  time = c(time, runtime[[3]]) 

  ## step 3
  start = proc.time()
  M = M %*% t(B)
  runtime = proc.time() - start
  time = c(time, runtime[[3]]) 

  ## step 4
  start = proc.time()
  D = E %*% t(B)
  runtime = proc.time() - start
  time = c(time, runtime[[3]]) 

  ## step 5
  start = proc.time()
  M = M/D
  runtime = proc.time() - start
  time = c(time, runtime[[3]]) 

  ## step 6
  start = proc.time()
  A = A*M
  runtime = proc.time() - start
  time = c(time, runtime[[3]]) 

  ## step 7
  start = proc.time()
  A <- pmax(A,e)
  runtime = proc.time() - start
  time = c(time, runtime[[3]]) 
  return(list(A = A, time = time))
}


update_B_timed <- function(X,A,B,E,e){
  # original update is:
  # B <- B * ((t(A) %*% (X / (A %*% B))) / (t(A) %*% E))
  # B <- pmax(B,e)
  time = c()

  ## step 1
  start = proc.time()
  M = A %*% B
  runtime = proc.time() - start
  time = c(time, runtime[[3]])

  ## step 2
  start = proc.time()
  M = X / M
  runtime = proc.time() - start
  time = c(time, runtime[[3]]) 

  ## step 3
  start = proc.time()
  M = t(A) %*% M
  runtime = proc.time() - start
  time = c(time, runtime[[3]]) 

  ## step 4
  start = proc.time()
  D = t(A) %*% E
  runtime = proc.time() - start
  time = c(time, runtime[[3]]) 

  ## step 5
  start = proc.time()
  M = M/D
  runtime = proc.time() - start
  time = c(time, runtime[[3]]) 

  ## step 6
  start = proc.time()
  B = B*M
  runtime = proc.time() - start
  time = c(time, runtime[[3]]) 

  ## step 7
  start = proc.time()
  B <- pmax(B,e)
  runtime = proc.time() - start
  time = c(time, runtime[[3]]) 
  return(list(B = B, time = time))
}






betanmf_timed <- function (X, A, B, numiter = 1000, e = .Machine$double.eps, verbose = TRUE) {
  if (inherits(X,"matrix"))
    X <- as.matrix(X)
  n <- nrow(X)
  m <- ncol(X)
  E <- matrix(1,n,m)
  # progress <- data.frame(iter = 1:numiter,objective = 0,
  #                        max.diff = 0,timing = 0)

  time_A = c()
  time_B = c()
  time_cost = c()
      
  # Repeat until we reach the number of requested iterations.
  if (verbose)
    cat("iter         objective max.diff\n")
  for (i in 1:numiter) {

    # Save the current estimates of the factors and loadings.
    A0 <- A
    B0 <- B

    

    # Update the loadings ("activations").
    #A <- A * (((X / (A %*% B)) %*% t(B)) / (E %*% t(B)))
    #A <- pmax(A,e)
    A_up <- update_A_timed(X,A,B,E,e)
    A <- A_up$A
    time_A = rbind(time_A, A_up$time)

    # Update the factors ("basis vectors").
    # B <- B * ((t(A) %*% (X / (A %*% B))) / (t(A) %*% E))
    # B <- pmax(B,e)
    B_up <- update_B_timed(X,A,B,E,e)
    B <- B_up$B
    time_B = rbind(time_B, B_up$time)
  

    # Compute the value of the objective (cost) function at the
    # current estimates of the factors and loadings.
    start = proc.time()
    f <- cost(X,A %*% B,e)
    time = proc.time() - start
    time_cost = rbind(time_cost, time[[3]])
    d <- max(max(abs(A - A0)),max(abs(B - B0)))
    # progress[i,"objective"] <- f
    # progress[i,"max.diff"]  <- d
    # progress[i,"timing"]    <- timing["elapsed"]
    if (verbose)
      cat(sprintf("%4d %0.10e %0.2e\n",i,f,d))
  }

  return(list(A = A,B = B,value = f,time_A = time_A, time_B = time_B, time_cost = time_cost))
}
